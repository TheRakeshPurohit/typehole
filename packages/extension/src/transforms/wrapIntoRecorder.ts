import * as ts from "typescript";

import { getAST } from "../parse/module";

function nodeToParameterExpression(node: ts.Node): ts.Expression {
  if (ts.isFunctionDeclaration(node)) {
    return ts.factory.createFunctionExpression(
      node.modifiers,
      node.asteriskToken,
      node.name,
      node.typeParameters,
      node.parameters,
      node.type,
      node.body!
    );
  }

  if (ts.isExpressionStatement(node)) {
    return node.expression;
  }
  return node as ts.Expression;
}

export function wrapIntoRecorder(id: number, expressionCode: string) {
  const ast = getAST(expressionCode);
  const sourceFile = ast.getSourceFile();

  let rootNode: ts.Node = ast;

  while (
    ts.isSourceFile(rootNode) ||
    ts.SyntaxKind[rootNode.kind] === "SyntaxList"
  ) {
    rootNode = rootNode.getChildAt(0);
  }

  const wrapped = ts.factory.createCallExpression(
    ts.factory.createPropertyAccessExpression(
      ts.factory.createIdentifier("typehole"),
      ts.factory.createIdentifier(`t${id === 0 ? "" : id}`)
    ),
    undefined,
    [nodeToParameterExpression(rootNode)]
  );

  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
  const result = printer.printNode(
    ts.EmitHint.Unspecified,
    wrapped,
    sourceFile
  );

  return result;
}
