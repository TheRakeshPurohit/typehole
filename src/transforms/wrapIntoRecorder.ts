import * as ts from "typescript";
import { tsquery } from "@phenomnomnominal/tsquery";

export function wrapIntoRecorder(id: number, expressionCode: string) {
  const ast = tsquery.ast(expressionCode);
  const sourceFile = ast.getSourceFile();
  const expression = ast.getChildAt(0).getChildAt(0).getChildAt(0);

  const wrapped = ts.factory.createCallExpression(
    ts.factory.createIdentifier(`typehole.t${id === 0 ? "" : id}`),
    undefined,
    [expression as ts.Expression]
  );

  const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
  const result = printer.printNode(
    ts.EmitHint.Unspecified,
    wrapped,
    sourceFile
  );
  return result;
}
