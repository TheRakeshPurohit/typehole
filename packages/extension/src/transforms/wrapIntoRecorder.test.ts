import { wrapIntoRecorder } from "./wrapIntoRecorder";

test("wraps function declaration into a recorder", () => {
  const selectedExpression = `
  function App() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <p>
            Edit <code>src/App.tsx</code> and save to{" "}
          </p>
          <a
            className="App-link"
            href="https://reactjs.org"
            target="_blank"
            rel="noopener noreferrer"
          >
            Learn React
          </a>
        </header>
      </div>
    );
  }
  `;
  const actual = wrapIntoRecorder(0, selectedExpression);
  expect(actual).toMatchSnapshot();
});

test("wraps expressions into recorder call", () => {
  const actual = wrapIntoRecorder(
    0,
    `
  tsquery.query(
    ast,
    "InterfaceDeclaration > Identifier[name='AutoDiscover']"
  )
  `
  );
  expect(actual).toMatchSnapshot();
});

test("wraps literals into recorder calls", () => {
  expect(wrapIntoRecorder(0, `1 + 3 + 4`)).toMatchSnapshot();
  expect(wrapIntoRecorder(0, `"moro" + "moro"`)).toMatchSnapshot();
  expect(wrapIntoRecorder(0, `["moro", "moro"]`)).toMatchSnapshot();
  expect(wrapIntoRecorder(0, `[() => 3, "moro"]`)).toMatchSnapshot();
});
