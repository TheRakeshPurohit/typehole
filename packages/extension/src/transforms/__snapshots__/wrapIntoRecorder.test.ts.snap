// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finds selected expression 1`] = `
"typehole.t(function App() {
    return (<div className=\\"App\\">
        <header className=\\"App-header\\">
          <img src={logo} className=\\"App-logo\\" alt=\\"logo\\"/>
          <p>
            Edit <code>src/App.tsx</code> and save to{\\" \\"}
          </p>
          <a className=\\"App-link\\" href=\\"https://reactjs.org\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
            Learn React
          </a>
        </header>
      </div>);
})"
`;

exports[`wraps expressions into recorder call 1`] = `"typehole.t(tsquery.query(ast, \\"InterfaceDeclaration > Identifier[name='AutoDiscover']\\"))"`;

exports[`wraps function declaration into a recorder 1`] = `
"typehole.t(function App() {
    return (<div className=\\"App\\">
        <header className=\\"App-header\\">
          <img src={logo} className=\\"App-logo\\" alt=\\"logo\\"/>
          <p>
            Edit <code>src/App.tsx</code> and save to{\\" \\"}
          </p>
          <a className=\\"App-link\\" href=\\"https://reactjs.org\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
            Learn React
          </a>
        </header>
      </div>);
})"
`;

exports[`wraps literals into recorder calls 1`] = `"typehole.t(1 + 3 + 4)"`;

exports[`wraps literals into recorder calls 2`] = `"typehole.t(\\"moro\\" + \\"moro\\")"`;

exports[`wraps literals into recorder calls 3`] = `"typehole.t([\\"moro\\", \\"moro\\"])"`;

exports[`wraps literals into recorder calls 4`] = `"typehole.t([() => 3, \\"moro\\"])"`;
